interpreter/syntax/ast.ml

```diff
@@ -140,18 +140,23 @@ and memory' =
   mtype : memory_type;
 }
 
-type ('data, 'ty) segment = ('data, 'ty) segment' Source.phrase
-and ('data, 'ty) segment' =
-  | Active of {index : var; offset : const; init : 'data}
-  | Passive of {etype : 'ty; data : 'data}
-
 type elem = elem' Source.phrase
 and elem' =
   | RefNull
   | RefFunc of var
 
-type table_segment = (elem list, elem_type) segment
-type memory_segment = (string, unit) segment
+
+type table_segment = table_segment' Source.phrase
+and table_segment' =
+  | ActiveWithRefs of {index : var; offset : const; etype : elem_type; init : elem list}
+  | ActiveWithIndices of {index : var; offset : const; init : elem list}
+  | PassiveWithRefs of {etype : elem_type; data : elem list}
+  | PassiveWithIndices of {data : elem list}
```

Comment:
AFAICS, there is no value in carrying this distinction into the AST, it just complicates everything downstream. Instead, simply decode the index cases as (a short-hand for) ref cases, converting the extern_kind into a suitable reftype. In the encoder, pick index form when possible.

If I'm not mistaken, that would also make the duplication of the segment types and code unnecessary, but I may be missing something.

Reply:
> The reason for this distinction is that it prevents an inconsistent state.

The elem_kind won't occur in the AST either, so there is no inconsistency. Simply think of the index forms as a shorthand in the binary format. The "desugaring" converts the elem_kind into an elem_type.

We want to keep the AST as simple and canonical as possible, so local optimisations of the binary format should not leak into it.

> About sharing code between element segments and data segments: It felt like the overlap was already limited

Yeah, but it's still nice to share it. Unless there are strong reasons for removing it I'd prefer not to.

> Question: why were there separate definitions for Active and Passive segments? Could they not also just be distinguished by a flag?

Well, the constructor is the flag. And the presence of some fields depends on it, which a Boolean flag couldn't express, not without allowing inconsistent state ;).